feat[next-dace]: Use SDFG library node for lowering of broadcast

edopao opened the pull request:
TODO:
- [x] Run ICON4Py CI, see [ICON4Py PR#1240](https://github.com/C2SM/icon4py/pull/1240)
- [x] Run Blueline (there are some degradation but in general it works)
- [ ] Run MuPhys (just to be sure)

edopao committed:
edit

edopao committed:
edit

edopao committed:
undo extra change

edopao committed:
use library node also in concat_where

philip-paul-mueller reviewed (commented):
There are some refinements needed.

philip-paul-mueller commented on src/gt4py/next/program_processors/runners/dace/sdfg_library_nodes.py:
```diff
@@ -0,0 +1,73 @@
+
+
+@dace_library.node
+class Fill(dace_nodes.LibraryNode):
```
I do not fully understand this comment.
Regarding the not materialization, this PR is not complete, because some transformation, to handle it are missing; Which will be in some other PR.
Currently they expand to a Map and then the optimizer has to guess what to do and currently the Maps are handled in a crude way.
The main reason to have this Node is that the optimizer (or some transformation, I think we need two) can handle it in a more appropriate way.

edopao replied:
I think I get Hannes' comment. After changing the library node to a mapped tasklet inlined in the same state (as suggested by Philip) we can achieve the goal suggested by Hannes. I have spotted some places in the lowering of `concat_where` where I could use the new broadcast node. For now I have just added a TODO.

edopao commented on src/gt4py/next/program_processors/runners/dace/gtir_to_sdfg_primitives.py:
```diff
@@ -73,7 +73,7 @@ def _parse_fieldop_arg(
     sdfg_builder: gtir_to_sdfg.SDFGBuilder,
     domain: gtir_domain.FieldopDomain,
-) -> MaybeNestedInTuple[gtir_dataflow.IteratorExpr | gtir_dataflow.MemletExpr]:
+) -> gtir_dataflow.IteratorExpr | gtir_dataflow.MemletExpr:
```
This is the correct return type. This function can never return a tuple. You are right that this change is unrelated.

edopao commented on src/gt4py/next/program_processors/runners/dace/sdfg_library_nodes.py:
```diff
@@ -0,0 +1,73 @@
+    environments: Final[list[Any]] = []
+
+    @staticmethod
+    def expansion(node: Fill, parent_state: dace.SDFGState, parent_sdfg: dace.SDFG) -> dace.SDFG:
```
Thanks for pointing me to this, I had missed it.

edopao committed:
edit

edopao committed:
Merge remote-tracking branch 'upstream/main' into dace-fill_node

edopao committed:
edit

edopao committed:
fix for inf expressions

edopao commented on src/gt4py/next/program_processors/runners/dace/gtir_to_sdfg_primitives.py:
```diff
@@ -285,6 +291,86 @@ def translate_as_fieldop(
+
+    # Retrieve the scalar argument, which could be either a literal value or the
+    # result of a scalar expression.
+    if isinstance(scalar_arg, gtir.Literal) and np.isfinite(
```
The filter `np.isfinite()` is currently just a workaround to a dace issue related to parsing of `inf` value in `test_execution::test_infinity`

edopao committed:
edit

edopao committed:
edit

edopao committed:
Merge branch 'dace-refactor_concat_where' into dace-fill_node

edopao committed:
edit

edopao committed:
Merge branch 'dace-refactor_concat_where' into dace-fill_node

edopao committed:
Merge branch 'main' into dace-fill_node

edopao committed:
Merge branch 'main' into dace-fill_node

edopao committed:
remove special handling for inf symbol

edopao committed:
fix rebase

edopao committed:
fix rebase

edopao committed:
Merge branch 'main' into dace-fill_node

edopao commented:
No plan for now to integrate this feature.

edopao committed:
Merge branch 'main' into dace-fill_node

edopao committed:
Merge branch 'main' into dace-fill_node

edopao committed:
edit

edopao committed:
add library_nodes module folder

edopao committed:
edit

edopao committed:
undo extra change

edopao committed:
Merge branch 'main' into dace-fill_node

edopao committed:
edit

edopao committed:
fix

edopao committed:
fix

philip-paul-mueller committed:
Started with the broadcast inline transformation.

philip-paul-mueller committed:
Fixed some wrong names.

philip-paul-mueller committed:
Fix

philip-paul-mueller committed:
In C++ Speak it compiles but the test does not pass.

philip-paul-mueller committed:
Fixed two errors.

philip-paul-mueller committed:
Added a test that ensures that indirect accessing is not triggered.

philip-paul-mueller committed:
Continued.

philip-paul-mueller committed:
Some more notes.

philip-paul-mueller committed:
I think I need to merge these two transformations.

philip-paul-mueller committed:
Merged it and now pre-commit is now happy, but does it works.

philip-paul-mueller committed:
Now at least the unit tests passes.

philip-paul-mueller committed:
Added a test for the access mode case.

philip-paul-mueller committed:
Added a test for the fan out of AccessNodes.

philip-paul-mueller committed:
Added multi connection test.

philip-paul-mueller committed:
Added more tests to the broadcast node.

philip-paul-mueller committed:
Started with an inline expansion, but the regular expansion needs to be updated yet.

philip-paul-mueller committed:
Updated the Broadcast expansion. Not integrated yet.

philip-paul-mueller committed:
Merge remote-tracking branch 'gt4py/main' into dace-fill_node_philip

philip-paul-mueller committed:
Fixed an error with the life time.

philip-paul-mueller committed:
Made a first integration, that is probably not correct.

philip-paul-mueller committed:
Made the inline expander accessible to the outside world.

philip-paul-mueller committed:
Fixed some issues in the validatiobn function of the broadcast node. Did I even run the tests?

philip-paul-mueller committed:
I do not fully understand why this is needed, sometimes NumPy is just confusing.

edopao committed:
Merge branch 'main' into dace-fill_node

philip-paul-mueller committed:
Fixed the expander, it should now work.

philip-paul-mueller committed:
Fixed the broadcasting test since `unique_name()` has been moved.

philip-paul-mueller committed:
Added a test for the library expansion, only scalar yet.

philip-paul-mueller committed:
Fixed an error in validation.

philip-paul-mueller committed:
Added a test to operate on vectors.

philip-paul-mueller committed:
NOw we also do some slicing.

philip-paul-mueller committed:
Update src/gt4py/next/program_processors/runners/dace/lowering/gtir_to_sdfg_primitives.py

Co-authored-by: Edoardo Paone <edoardo16@gmail.com>

philip-paul-mueller committed:
Discussion points with Edoardo.

philip-paul-mueller committed:
Fixed a wrong annotation.

philip-paul-mueller committed:
Fixed some issue with renaming.

philip-paul-mueller committed:
Something is not working as expected.

philip-paul-mueller committed:
Refined how the `MapToCopy` detects broadcasts.

philip-paul-mueller committed:
Handled how expansion of librarynodes is currently performed.

philip-paul-mueller committed:
Fixed extensive validation in `FuseHorizontalConditionBlocks`.

philip-paul-mueller committed:
Integrated the correction.

philip-paul-mueller committed:
Incorporated the domain correction at the expansion.

philip-paul-mueller committed:
Made a unit test a bit harder.

philip-paul-mueller committed:
Makeing the optimnizer temporaraly more strict.

philip-paul-mueller committed:
Merge branch 'dace-fill_node_philip' into dace-fill_node

philip-paul-mueller committed:
Now also the parameter names are correctly set.

philip-paul-mueller committed:
Implemented a missing case.

philip-paul-mueller committed:
Forgot to update them.

philip-paul-mueller committed:
git Fixed a wrong check.

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+    library as dace_library,
+    nodes as dace_nodes,
+    properties as dace_properties,
+    subsets as dace_sbs,
```
In the transformation module, we use the `dace_sbs` alias, in the lowering module we use `dace_subsets`. It's OK to use `dace_sbs` in this module, but let's try to keep it consistent.

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
````diff
@@ -0,0 +1,302 @@
+    ```python
+    for i in range(len(broadcast_in_dim):
+        assert output.shape[broadcast_in_dim[i]] == value_to_broadcast.shape[i]
+    ```
````
```suggestion
    ```
    In other words, the result array shape has the same size as the broadcast domain.
```

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
````diff
@@ -0,0 +1,302 @@
+    ```
+
+    Args:
+        broadcast_in_dim: How to broadcast.
````
```suggestion
        broadcast_in_dim: How to broadcast, see the class documentation.
```

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+
+    Args:
+        broadcast_in_dim: How to broadcast.
+        params: The parameters that should be used for the expansion. If given one
```
```suggestion
        params: The parameters that should be used for the expansion. If given, one
```

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+
+    Todo:
+        - While for the output it is probably okay to always require an adjacent
+            AccessNode for the input it might be possible to be on the other side
```
```suggestion
            AccessNode, the input nodes might be outside a map scope.
```
However, I don't understand how this could happen.

philip-paul-mueller replied:
I think this comment dates back when there was still the `.value` attribute.

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+        self,
+        name: str,
+        broadcast_in_dims: Sequence[int],
+        params: Optional[Iterable[gtx_common.Dimension | str]],
```
```suggestion
        params: Iterable[gtx_common.Dimension | str] | None,
```

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+        debuginfo: dace.dtypes.DebugInfo | None = None,
+    ):
+        # TODO(philip, edopao): I would propose to drop `value` then. This makes it
+        #   simpler to handle in the transformations.
```
I agree, we can remove the todo comment.

```suggestion
```

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+
+    def validate(self, sdfg: dace.SDFG, state: dace.SDFGState) -> None:
+        if len(self.brodcast_in_dims) != len(set(self.brodcast_in_dims)):
+            raise ValueError("`Can not broadcast to multiple dimensions at the same time.")
```
```suggestion
            raise ValueError("`Broadcast dimensions must be unique")
```

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+            raise ValueError("`Can not broadcast to multiple dimensions at the same time.")
+
+        # TODO(phimuell): Handle empty Memlets in the input.
+        if state.in_degree(self) != 1 and next(iter(state.in_edges(self))).dst_conn == _INPUT_NAME:
```
```suggestion
        if state.in_degree(self) != 1 or next(iter(state.in_edges(self))).dst_conn != _INPUT_NAME:
```

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+            raise ValueError("GT4Py Broadcast node needs exactly one input.")
+        if (
+            state.out_degree(self) != 1
+            and next(iter(state.out_edges(self))).src_conn == _OUTPUT_NAME
```
```suggestion
            state.out_degree(self) != 1
            or next(iter(state.out_edges(self))).src_conn != _OUTPUT_NAME
```

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+        if (self.params is not None) and (len(self.params) != len(bcast_result_desc.shape)):
+            raise ValueError(
+                f"Expected that {len(bcast_result_desc.shape)} parameters are"
+                f" needed but {len(self.params)} were specified."
```
`The number of dimensions of the result shape {len(bcast_result_desc.shape)} does not match the nmuber of broadcast parameters {len(self.params)}.`

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+        if isinstance(bcast_value_desc, dace_data.Scalar):
+            if len(self.brodcast_in_dims) != 0:
+                raise ValueError("For a scalar `broadcast_in_dims` must be empty.")
+        else:
+            if len(self.brodcast_in_dims) != len(bcast_value_desc.shape):
```
```suggestion
        elif len(self.brodcast_in_dims) != len(bcast_value_desc.shape):
```

philip-paul-mueller replied:
I think the way it is now is a bit clearer because it makes it clear that "this is the Array case and there is the Scalar checks".

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+
+            for src_dim, bcast_dst_dim in enumerate(self.brodcast_in_dims):
+                if bcast_dst_dim < 0:
+                    raise ValueError("Negative broadcast")
```
```suggestion
                    raise ValueError("Negative dimension index is invalid.")
```

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+        iter(state.out_edges(bcast_node))
+    )
+
+    # TODO(phimuell): Add warning.
```
I don't think we should add a warning. Either allow it or not.

philip-paul-mueller replied:
See my updated comment.

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+    output_subset: list[str] = map_params.copy()
+    map_ranges: dict[str, dace_sbs.Range] = {
+        map_param: dace_sbs.Range([sbs])
+        for map_param, sbs in zip(map_params, output_edge.data.subset)
```
```suggestion
        for map_param, sbs in zip(map_params, output_edge.data.subset, strict=True)
```

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+        bcast_value_offset = input_edge.data.subset.min_element()
+        input_subset = [
+            f"{map_params[dst_dim]} + ({offset})"
+            for dst_dim, offset in zip(bcast_node.brodcast_in_dims, bcast_value_offset)
```
```suggestion
            for dst_dim, offset in zip(bcast_node.brodcast_in_dims, bcast_value_offset, strict=True)
```

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+        code="__out = __in",
+    )
+
+    state.add_edge(
+        input_edge.src,
+        input_edge.src_conn,
+        me,
+        f"IN_{input_edge.data.data}",
+        dace.Memlet.from_array(input_edge.data.data, sdfg.arrays[input_edge.data.data]),
+    )
+    state.add_edge(
+        me,
+        f"OUT_{input_edge.data.data}",
+        bcast_tlet,
+        "__in",
+        dace.Memlet(data=input_edge.data.data, subset=", ".join(input_subset)),
+    )
+    me.add_scope_connectors(input_edge.data.data)
```
```suggestion
    state.add_memlet_path(
        input_edge.src,
        me,
        bcast_tlet,
        src_conn=input_edge.src_conn,
        dst_conn="__in",
        dace.Memlet(data=input_edge.data.data, subset=", ".join(input_subset)),
    )
```

philip-paul-mueller replied:
I am actually not a fan of `add_memlet_path()`.

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+        output_edge.dst_conn,
+        dace.Memlet(data=output_edge.data.data, subset=copy.deepcopy(output_edge.data.subset)),
+    )
+    mx.add_scope_connectors(output_edge.data.data)
```
```suggestion
    state.add_memlet_path(
        bcast_tlet,
        mx,
        output_edge.dst,
        src_conn="__out",
        dst_conn=output_edge.dst_conn,
        dace.Memlet(data=output_edge.data.data, subset=copy.deepcopy(output_edge.data.subset)),
    )
```

edopao commented on src/gt4py/next/program_processors/runners/dace/library_nodes/broadcast.py:
```diff
@@ -0,0 +1,302 @@
+    def expansion(node: Broadcast, state: dace.SDFGState, sdfg: dace.SDFG) -> dace.SDFG:
+        # TODO:
+        #   - Modify the edges on the outside.
+        #   - Handle the missing symbols.
```
The missing symbols are automatically mapped to symbols with the same name, aren't they?

philip-paul-mueller replied:
Yes, this thing was also very old. You can see the note at the end of the function.

edopao commented on src/gt4py/next/program_processors/runners/dace/transformations/broadcast.py:
```diff
@@ -0,0 +1,422 @@
+
+
+@dace_properties.make_properties
+class ScalarBrodcastInliner(dace_transformation.SingleStateTransformation):
```
```suggestion
class ScalarBroadcastInliner(dace_transformation.SingleStateTransformation):
```

edopao commented on src/gt4py/next/program_processors/runners/dace/lowering/gtir_dataflow.py:
```diff
@@ -27,14 +27,18 @@
 from gt4py.next.iterator.transforms import symbol_ref_utils
-from gt4py.next.program_processors.runners.dace import sdfg_args as gtx_dace_args
+from gt4py.next.program_processors.runners.dace import (
+    library_nodes as gtir_library_nodes,
```
```suggestion
    library_nodes as gtx_dace_library_nodes,
```

edopao commented on src/gt4py/next/program_processors/runners/dace/lowering/gtir_dataflow.py:
```diff
@@ -1471,27 +1342,73 @@ def _visit_reduce(self, node: gtir.FunCall) -> ValueExpr:
+            )
+            reduce_node.input_conn = inp_conn
+            reduce_node.output_conn = outp_conn
+            reduce_node.mask_conn = mask_conn
```
```suggestion
            reduce_node = gtir_library_nodes.ReduceWithSkipValues(
                name,
                reduce_wcr,
                identity=reduce_identity.value,
                init=reduce_init.value,
                debuginfo=gtir_to_sdfg_utils.debug_info(node),
                input_conn=inp_conn,
                output_conn=outp_conn,
                mask_conn=mask_conn,
            )
```

edopao commented on src/gt4py/next/program_processors/runners/dace/lowering/gtir_to_sdfg_primitives.py:
```diff
@@ -292,6 +290,162 @@ def translate_as_fieldop(
+    fieldop_expr, fieldop_domain_expr = fun_node.args
+    assert cpm.is_ref_to(fieldop_expr, "deref")
+
+    # TODO:
+    #   - Include the dimensions of the output, such that we can generate the Map correctly
+    #       this is needed for expansion of the broadcast node such that we can order them correctly.
+    #       This information is stored in `field_dims`.
```
Is this TODO still valid?

edopao commented on src/gt4py/next/program_processors/runners/dace/lowering/gtir_to_sdfg_primitives.py:
```diff
@@ -292,6 +290,162 @@ def translate_as_fieldop(
+    if isinstance(bcast_arg, gtir.Literal):
+        assert isinstance(bcast_arg.type, ts.ScalarType)
+        bcast_value_tlet, connector_mapping = sdfg_builder.add_tasklet(
+            sdfg_builder.unique_tasklet_name(bcast_node_name),
```
The name is made unique inside the `add_tasklet()` wrapper:
```suggestion
            bcast_node_name,
```

edopao commented on src/gt4py/next/program_processors/runners/dace/lowering/gtir_to_sdfg_primitives.py:
```diff
@@ -292,6 +290,162 @@ def translate_as_fieldop(
+            connector_mapping["__out"],
+            bcast_value,
+            None,
+            dace.Memlet.from_array(bcast_value_name, bcast_value_desc),
```
```suggestion
            dace.Memlet(data=bcast_value_name, subset="0"),
```

edopao commented on src/gt4py/next/program_processors/runners/dace/lowering/gtir_to_sdfg_primitives.py:
```diff
@@ -292,6 +290,162 @@ def translate_as_fieldop(
+
+        bcast_value = arg.field
+        bcast_value_desc = bcast_value.desc(ctx.sdfg)
+        if isinstance(bcast_value_desc, dace_data.Scalar):
+            broadcast_in_dims = []
```
Have you ever hit this case: `IteratorExpr` with `bcast_value_desc` scalar? I think you can remove this if branch.

philip-paul-mueller replied:
I think I hit it.

edopao commented on src/gt4py/next/program_processors/runners/dace/lowering/gtir_to_sdfg_primitives.py:
```diff
@@ -292,6 +290,162 @@ def translate_as_fieldop(
+    )
+    ctx.state.add_node(bcast_node)
+
+    # If not specified differently use the whole `bcast_value` content.
+    if bcast_value_subset is None:
+        bcast_value_subset = dace_subsets.Range.from_array(bcast_value.desc(ctx.sdfg))
+
```
```suggestion
    # If not specified differently use the whole `bcast_value` content.
    if bcast_value_subset is None:
        assert isinstance(bcast_value_desc, dace_data.Scalar)
        bcast_value_subset = dace_subsets.Range.from_string("0")

```

philip-paul-mueller replied:
I would rather keep it in this way, because it will work also in the future.

edopao commented on src/gt4py/next/program_processors/runners/dace/transformations/auto_optimize.py:
```diff
@@ -247,6 +252,15 @@ def gt_auto_optimize(
+                gtir_library_nodes.inplace_broadcast_expander(node, state, state.sdfg)
+            elif isinstance(node, gtir_library_nodes.GTIR_LIBRARY_NODES):
+                node.expand(state)
+
```
```suggestion
        nb_applied = 0
        for node, state in sdfg.all_nodes_recursive():
            if isinstance(node, gtir_library_nodes.Broadcast):
                gtir_library_nodes.inplace_broadcast_expander(node, state, state.sdfg)
                nb_applied += 1
            elif isinstance(node, gtir_library_nodes.GTIR_LIBRARY_NODES):
                node.expand(state)
                nb_applied += 1
        if validate_all and nb_applied > 0:
            sdfg.validate()
```

philip-paul-mueller replied:
Good idea.

edopao commented on src/gt4py/next/program_processors/runners/dace/transformations/broadcast.py:
```diff
@@ -0,0 +1,422 @@
+    broadcast library nodes. However, this transformations only operate on
+    scalar values that are broadcasted.
+
+    The first one is "static access replacement" in that operation the pattern:
```
```suggestion
    The transformation performs two kinds of operations that are related to
    broadcast library nodes. However, this transformation only operates on
    broadcast of scalar values.

    The first one is "static access replacement" in that the operation pattern:
```

edopao commented on src/gt4py/next/program_processors/runners/dace/transformations/broadcast.py:
````diff
@@ -0,0 +1,422 @@
+
+    The second kind of operation is "AccessNode bypassing" in which the
+    following pattern is matched:
+    ```
+        (value_to_broadcast) -> (bcast_result) -> (access_node_consumer)
+    ```
+    to:
+    ```
+        (value_to_broadcast) -> (access_node_consumer)
+    ```
````
```suggestion
    ```
        (value_to_broadcast)  -> [BroadCast] -> (bcast_result) -> (access_node_consumer)
    ```
    to:
    ```
        (value_to_broadcast) -> [BroadCast] -> (access_node_consumer)
    ```
```

edopao commented on src/gt4py/next/program_processors/runners/dace/transformations/broadcast.py:
```diff
@@ -0,0 +1,422 @@
+    Thus, the broadcast is performed directly into the target destination.
+
+    It is important that these two modes can be performed at the same time.
+    Meaning `bcast_result` can have multiple consumer.
```
```suggestion
    Meaning `bcast_result` can have multiple consumers.
```

edopao commented on src/gt4py/next/program_processors/runners/dace/transformations/broadcast.py:
```diff
@@ -0,0 +1,422 @@
+
+from gt4py.next.program_processors.runners.dace import (
+    library_nodes as gtx_lib_nodes,
+    transformations as gtx_xtrans,
```
It is the first time I see this alias `gtx_xtrans`. As for the other, we can agree on a consistent alias across files (I have seen another alias above in this PR).

edopao commented on src/gt4py/next/program_processors/runners/dace/transformations/broadcast.py:
```diff
@@ -0,0 +1,422 @@
+        # A fundamental requirement is that `bcast_result` is only generated by us.
+        #  ADR-18 guarantees us this if it is transient and has a single producer,
+        #  `bcast_node`. However, since we will remove `bcast_result`, we have to
+        #  make sure that it is not used every where else.
```
```suggestion
        #  make sure that it is not used anywhere else.
```

edopao commented on src/gt4py/next/program_processors/runners/dace/transformations/broadcast.py:
```diff
@@ -0,0 +1,422 @@
+
+            match consumer := consumer_edge.dst:
+                case dace_nodes.AccessNode():
+                    # TODO(phimuell): Are there more checks needed.
```
I suggest removing this todo comment before merge, unless there are known cases.
```suggestion
```

edopao commented on src/gt4py/next/program_processors/runners/dace/transformations/broadcast.py:
```diff
@@ -0,0 +1,422 @@
+        # Check single use data if it was not known at the beginning.
+        if self._single_use_data is None:
+            find_single_use_data = dace_analysis.FindSingleUseData()
+            single_use_data = find_single_use_data.apply_pass(sdfg, None)
```
Would it be wrong to now store `single_use_data`? I am asking because it is used again inside `apply()`.

philip-paul-mueller replied:
No because, there is no guarantee that between now and later there is a change that makes data no longer single use.
By passing it on constructor the user kinds of guarantee that it is safe.

edopao commented on src/gt4py/next/program_processors/runners/dace/transformations/auto_optimize.py:
```diff
@@ -247,6 +252,15 @@ def gt_auto_optimize(
 
+        # We need new transformations in order to deal with GTIR library nodes.
+        # For now, we simply expand these nodes before starting optimizing.
+        # TODO: Remove once transformations are ready.
```
Why not calling `ScalarBrodcastInliner` before expanding?

edopao commented on src/gt4py/next/program_processors/runners/dace/transformations/broadcast.py:
```diff
@@ -0,0 +1,422 @@
+        #   probably yes, as we can remove the read and write of the initial data
+        #   only the write to final destination is left. If the consumers are Maps
+        #   the thing is a bit different. As we have to create the intermediate
+        #   allocation. If the read of the memory is okay the `InlineBroadcastAccess`
```
`InlineBroadcastAccess` does  not exist yet.

philip-paul-mueller replied:
I think that was the old name.
Then I removed it to `Scalar...`, because I thought that I onlyneed to handle scalars which was wrong.
So there is a todo to rename it, I think I will now switch to the old name.

edopao reviewed (commented):
Very good, just some minor comments.

edopao reviewed (commented):
Very good, just some minor comments.

philip-paul-mueller committed:
Updated the transformation to also handle vectors.

No unit test yet and also it has thr wrong name.

philip-paul-mueller committed:
Moved the `ScalarBroadcastInliner` (which still has the wrong name) into simplify.

Also expanding is now happening at the very end.

philip-paul-mueller committed:
Merge remote-tracking branch 'gt4py/main' into dace-fill_node

philip-paul-mueller committed:
Updated the `reconfigure_dataflow_after_rerouting()` such that it can partially handle additional domain.

philip-paul-mueller committed:
Added a note.

philip-paul-mueller committed:
Relocated the expansion of broadcast nodes.

They are now conditionally expanded in the auto optimizer if splitting is enabled and no broadcast related transformation has been applied.

philip-paul-mueller committed:
Added a transformation that can remove chains of broadcast operations.

philip-paul-mueller committed:
Updated the broadcast chain transformation to be more liberal.

philip-paul-mueller committed:
Integrated the broadcast chain remover into simplify.

philip-paul-mueller committed:
Applied Edoardo's suggestions.

philip-paul-mueller committed:
Removed name duplication when the broadcast nodes get dublicated in a transformation.

philip-paul-mueller committed:
Updated the broadcast library node

- Added a custom `free_symbols()` function to the library node to ensure that the symbols in `params` are not found
- Added validation to ensure that the name and lable attribute of the node are always in sync.

philip-paul-mueller committed:
Updated how expansion is working, should be more uniform in this way.

philip-paul-mueller committed:
This is empirically better, but still not good.

I assume that some map (horizontal) fusion transformations runs amock.
Probably there needs to be a better place.
Before I was thinking that the issue is with the fusion that the splitter performs.
But now I think it is earlier, i.e. Parallel Mapfusion has laready killed them.
Maybe it is a good idea to make the transformation split, i.e. map inlining is distinct from AccessNode replacement.

philip-paul-mueller committed:
Fixed how the reconnection is done.

philip-paul-mueller committed:
Splitting tools are now able to handle the broadcast node.

philip-paul-mueller committed:
I have seen a lot of strange thing in DaCe related to `other_subset` but this sets a new level.

philip-paul-mueller committed:
Simplify runs the broadcast transformations again.

philip-paul-mueller committed:
Updated the top level map optimizer.

Now the expansion of teh broadcast node takes place after the first fix point was found.
I think this is the best idea, but I am not fully sure.

philip-paul-mueller commented:
Here are the newest data from `8edefc0` which show that the speed is now comparable (at least is much better than [before](https://github.com/C2SM/icon4py/pull/1240#issuecomment-4398132484)).
If I look at my experiences with the `compute_advection_in_vertical` then the 3% degradation we observes are most likely caused by some broadcast loops that were (for whatever reasons) not integrated into other kernels.

<img width="670" height="1666" alt="bench_blueline_stencil_compute" src="https://github.com/user-attachments/assets/f26361ef-b4db-4425-9ca5-8610c217ea13" />

philip-paul-mueller committed:
This was too much validation.

philip-paul-mueller commented:
My current guess of why they are not integrated is because of their range.
In this PR the ranges are based on the size of the target location, i.e. the range is `[0, size_of_patch_to_broadcast)`, but before the range was based on the grid coordinates that were written, i.e. the range was `[136, 2000)`.
Furthermore, the Map splitter currently only consider the range and not the sizes.

I think the current design (with size) is better, because it makes handling the nodes much simpler and they contain less state (ideally they would be stateless, but this is not possible).
Thus, we should update the splitting tools to consider the size of the iteration spaces instead of their size.

philip-paul-mueller committed:
Merge branch 'main' into dace-fill_node

edopao commented:
I have moved the library node for reduction with skip values to a separate PR #2603

edopao committed:
Merge branch 'main' into dace-fill_node
